### PR TITLE
[WIP] quantized gromov wasserstein solver

### PR DESCRIPTION
**Pull request by cedricvincentcuaz**
## Types of changes

Implementation of first solvers for the quantized Fused Gromov-Wasserstein (qFGW) distance.
- creation of new files `ot/gromov/_quantized.py` and `test/gromov/test_quantized.py` .
- `quantized_fused_gromov_wasserstein_partitioned` : main function to compute the OT between two partitioned joint spaces with minimal inputs to compute global and local alignments, so that the user can perform any partitioning and representant selection as pre-processing. A boolean option `build_OT` allows the user to construct the OT matrix between non-partitioned spaces which might require a lot of memory. Other inputs mainly relate to inner `fused_gromov_wasserstein`, `gromov_wasserstein` and `emd_1d` solvers.

remark 1 : i) only the (F)GW conditional gradient solver is considered here. Note that authors also made use of the entropic  projected gradient solver. ii) only the square loss is considered for both types of OT problems.
 
- `quantized_fused_gromov_wasserstein`: main function to compute the OT between two joint spaces with axioms used by authors to partition both spaces. Taking as inputs structure matrices `C1` and `C2` treated as graphs, optionally as feature matrices `F1` and `F2`, optionally other structure matrices `C1_aux` and `C2_aux` used for partitioning and representant selection implemented in `get_graph_partition` and `get_graph_representants`.

- `quantized_fused_gromov_wasserstein_samples`: main function to compute the OT between two distributions in different spaces potentially endowed with features, with axioms used by authors to partition both spaces. Taking as inputs samples `X1` and `X2` endowed with an euclidean geometry on their respective space, optionally with feature matrices `F1` and `F2`, where both can be considered for partitioning and representant selection (e.g kmeans) implemented in `get_partition_and_representants_samples`.

## How has this been tested (if it applies)

- tests for all configurations in  `test.test_gromov.py::test_quantized_gromov`

## PR checklist

- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: ot/gromov/_quantized.py [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+        Convergence information for inner problems and qGW loss.
+
+    References
+    ----------
+    .. [66] Chowdhury, S., Miller, D., & Needham, T. (2021).
+        Quantized gromov-wasserstein. ECML PKDD 2021. Springer International Publishing.
+
+    """
+    if (part_method in ['fluid', 'louvain'] or rep_method == 'pagerank') and (not networkx_import):
+        warnings.warn(
```
[the remark this answers is not in the document]
Comment (rflamary) [reply ot/gromov/_quantized.py:370]: this one cna be tested with the test-minila that do not install networkx

**Review by rflamary — COMMENTED** [review]
This is nice, I have a few comments below, contact me on slack if you want to discuss them

File: ot/gromov/_quantized.py [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+
+def _get_partition(C, npart, part_method='random', random_state=0, nx=None):
+    """
+    Partitioning a given structure matrix either 'random', or using one
+    of {'louvain', 'fluid'} algorithms from networkx, or {'spectral', 'kmeans'}
+    clustering from scikit-learn.
+
+    Parameters
+    ----------
+    C : array-like, shape (n, n) or (n, d)
```
[the remark this answers is not in the document]
Comment (cedricvincentcuaz) [reply ot/gromov/_quantized.py:43]: changed to `get_graph_partition` and `get_partition_and_representants_samples` to handle both settings separately.

File: ot/gromov/_quantized.py [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+    return nx.from_numpy(part, type_as=C0)
+
+
+def _get_representants(C, part, rep_method='random', random_state=0, nx=None):
+    """
+    Get representants for each partition of a given structure matrix either.
+
+    Parameters
+    ----------
+    C : array-like, shape (n, n) or (n, d)
```
[the remark this answers is not in the document]
Comment (cedricvincentcuaz) [reply ot/gromov/_quantized.py:136]: changed to `get_graph_representants` and `get_partition_and_representants_samples` to handle both settings separately.

**Pull request merged**